test/runnable/test15862.d

```diff
@@ -0,0 +1,51 @@
+// https://issues.dlang.org/show_bug.cgi?id=15862
+
+/*
+PERMUTE_ARGS:
+REQUIRED_ARGS: -O -release
+*/
+
+
+int* p() pure nothrow {return new int;}
+int[] a() pure nothrow {return [0];}
+Object o() pure nothrow {return new Object;}
+
+int* pn() pure nothrow @nogc {return null;}
+int[] an() pure nothrow @nogc {return null;}
+Object on() pure nothrow @nogc {return null;}
+
+void main()
+{
+    {
+        int* p1 = p();
+        int* p2 = p();
+
+        if (p1 is p2) assert(0);
```

Comment:
To get those asserts to be ran in `release`.
It's trying to solve a common problem with assert: https://forum.dlang.org/thread/ed318016-16e9-965f-4d66-c0c94dd2564c@sociomantic.com
I'd say lets revisit when we fix said issue.